Adaptive lighting

zegl opened the pull request:
Continuation of #60 to add support for Adaptive Lighting (#30).

* Rebased #60 on top of master
* Made Adaptive Lighting opt-in

One oddity that I've found is that HomeKit does not seem to respect disabling adaptive lighting for a light once it has been enabled the first time, it's still unclear if there is anything I can do to support this.

### To Do

- [ ] Add automated tests
- [x] Update CHANGELOG
- [x] Update documentation
- [x] Update documentation generator

itavero commented:
I haven't reviewed everything yet. I'll hope I have time for that soon.

One thing I did already notice: the adaptive lighting configuration should probably be a converter specific configuration as was introduced in #458.

macfr76 commented:
Quite enthousiat about this new feature ! Let me know how to try when it will ne merged !

itavero commented:
@zegl Did you see my previous comment?

Also curious if you have an update with regards to:
> One oddity that I've found is that HomeKit does not seem to respect disabling adaptive lighting for a light once it has been enabled the first time, it's still unclear if there is anything I can do to support this.

zegl commented:
@itavero Thanks for your ping!

> the adaptive lighting configuration should probably be a converter specific configuration

Yes, I looked a bit into this but didn't code anything. It makes sense and seems to be a straightforward change to make. I'll get around to work on it soon. :-)

I have no new knowledge with regards to the disabling of adaptive lightning. My first thought was that Apple might not support it too well. I'm guessing that it's rare for non-HomeBridge devices to drop support for adaptive lighting.

lcsjandrey commented:
Any updates? Still waiting. Would love see this happening before the holidays

zegl commented:
Hey. I've now rebased and reworked this PR to incorporate the earlier feedback. Would appreciate another review and hear your thoughts.

Changes since 283d5e75c70563e2816b6430701f954735e5e569 (last push):

* Rebased on top of v1.9.2
* Changed the configuration to be a "converter" configuration
* Removed the documentation generation changes. If this gets merged, we can work on update the generated device documentation afterwards.
* Removed the configuration for using transitions (which caused light "flickering" on my Hue lights) and the minimum temperature changes (unclear how necessary this limitation was). If these configurations are needed for some setups, they can always be added again later.
* Removed the code to prevent enabling adaptive lighting if using Homebridge versions earlier than v1.3.0-beta.46. (Adaptive Lighting is disabled by default, so this should not be a problem).

itavero commented:
Thanks again for your work on this. I'll try to have another peek in the coming days.

itavero commented on src/converters/light.ts:
```diff
@@ -29,7 +29,21 @@ import {
 import { EXP_COLOR_MODE } from '../experimental';
 
+interface LightConfig {
+  adaptive_lightning_enabled?: boolean;
```
It has a typo and I would remove the `_enabled` suffix.
If we need more features in the future, we can possibly make this into an `boolean | object` instead of just a boolean.

```suggestion
  adaptive_lighting?: boolean;
```

itavero commented on src/converters/light.ts:
```diff
@@ -58,6 +87,7 @@ class LightHandler implements ServiceHandler {
 
   public static readonly KEY_COLOR_MODE = 'color_mode';
   public static readonly COLOR_MODE_TEMPERATURE = 'color_temp';
+  // private static readonly MQTT_PROPERTY_TRANSITION = 'transition';
```
Commented code can be removed.

itavero commented on src/converters/light.ts:
```diff
@@ -390,6 +460,38 @@ class LightHandler implements ServiceHandler {
+          this.accessory.log.debug(
+            `Adaptive Lighting: Color temperature ${value} skipped for ${this.accessory.displayName}. ` +
+              `Previous: ${this.lastAdaptiveLightingTemperature}`
+          );
```
Maybe we should leave this log statement out. I don't think it adds any information, considering the check above.
Have you checked if the the check (ignore if `change < 1` ) is even needed?

itavero replied:
Ran a short test on my home setup. The check is needed, otherwise it will send the same value every minute or so (depending on how many lights you have in your house that might add quite some _unnecessary_ traffic).

I decided to leave the check and log statement in for now. Might help debug issues in the future.

itavero commented on src/converters/light.ts:
```diff
@@ -390,6 +460,38 @@ class LightHandler implements ServiceHandler {
+          return false;
+        }
+
+        this.accessory.log.info(`Adaptive Lightning: Update for ${this.accessory.displayName} temperature=${value}`);
```
Perhaps move this to the `debug` level for now. Might remove it later on depending on how much noise it generates.

itavero replied:
Changed to debug level.

itavero reviewed (commented):
Added a few comments. I'll also try to do a diff compared to the current master branch as well.

Furthermore, the tests for `light.ts` should be extended.
I might have some time next weekend/Monday to do so.

itavero commented:
Took me a bit longer to get to this, but at least of got the easy To Do items taken care of.
I'll start looking into the remarks I've added earlier now, as well as the automated tests.

Again, no promises on when it will be finished, my 3 month old son is kinda the boss of my spare time now. 👶

itavero commented:
The [documentation for AdaptiveLightingController](https://developers.homebridge.io/HAP-NodeJS/classes/AdaptiveLightingController.html) mentions:
> Detecting changes on the lightbulb side: Any manual change to ColorTemperature or Hue/Saturation is considered as a signal to turn AdaptiveLighting off. In order to notify the AdaptiveLightingController of such an event happening OUTSIDE of HomeKit you must call [disableAdaptiveLighting](https://developers.homebridge.io/HAP-NodeJS/classes/AdaptiveLightingController.html#disableAdaptiveLighting) manually!

This is currently not implemented. I'll see if there is a good way to detect this.

Additionally, it also mentions:
> When using Hue/Saturation: When using Hue/Saturation in combination with the ColorTemperature characteristic you need to update the respective other in a particular way depending on if being in "color mode" or "color temperature mode". When a write happens to Hue/Saturation characteristic in is advised to set the internal value of the ColorTemperature to the minimal (NOT RAISING an event). When a write happens to the ColorTemperature characteristic just MUST convert to a proper representation in hue and saturation values, with RAISING an event. As noted above you MUST NOT call the [setValue](https://developers.homebridge.io/HAP-NodeJS/classes/Characteristic.html#setValue) method for this, as this will be considered a write to the characteristic and will turn off AdaptiveLighting. Instead, you should use [updateValue](https://developers.homebridge.io/HAP-NodeJS/classes/Characteristic.html#updateValue) for this. You can and SHOULD use the supplied utility method [colorTemperatureToHueAndSaturation](https://developers.homebridge.io/HAP-NodeJS/classes/ColorUtils.html#colorTemperatureToHueAndSaturation) for converting mired to hue and saturation values.

This is also not implemented yet.

itavero commented:
Added some code that should handle the first point mentioned (regarding disabling Adaptive Lighting when an external change is detected).

For the second point mentioned, I think this is similar to what is also done in the experimental feature `COLOR_MODE`.

itavero commented:
I'll merge this for now, so I can make a prerelease. Based on the feedback from the community, we might have to make improvements in the future.

zegl commented:
Wow, huge! Thanks for working on this and pushing it over the finish line! ❤️ 

I'll update my installation from the homebaked version built on an earlier version of this PR to the latest release once ready.

zegl commented:
And congrats on the baby!

itavero commented:
Please add any feedback you might have to #590
